### PR TITLE
Type option interfaces

### PR DESCRIPTION
**Pull request by ryanio**
This PR adds the remaining typed option interfaces for class constructors.

### CODE REVIEW

**Comment by holgerd77** [comment]
@ryanio don't take any care of #160, I think this will likely be closed at some PIT without merge, I also overcame a lot of the changes there with subsequent work on #161. If it makes sense you can just copy-paste some parts over.

**Comment by ryanio** [comment]
I'm not sure why I got a bunch of `@typescript-eslint/no-unnecessary-condition` lint errors all of a sudden, but I resolved all of them. Some with replacing `||` for `??` which looks good to me since we now [`prefer-nullish-coalescing`](https://github.com/ethereumjs/ethereumjs-config/commit/f18efabd797f7bac8b87ee2dad68038ef5ba3489).

**Comment by ryanio** [comment]
On hold until #179 is merged, then I will rebase here.

**Comment by ryanio** [comment]
This is now ready for review and merge.

**Comment by ryanio** [comment]
There are still a good amount of TODOs, skipped tests, and `any` types that I aim to resolve in future PRs.

File: lib/net/server/server.ts
Comment (holgerd77) [inline lib/net/server/server.ts:42]: I made this a conscious decision after this singleton discussion with @rumkin and others to not do a default instantiation of `Config` within the classes - I remember that I took some significant extra round on adopting the tests after that - since this would introduce the danger to have an inconsistent config state in the code base if passing in the central config object would be forgotten (which would not be detected with this changed setup).

I would suggest that we stick with that - at least if you just reintroduced for some general convenience - or did you have explicit reasons to introduce this here?

Will not make this a blocker here though due to the size of the PR, think it is better if we got this quickly merged after the latest problems with conflicting work states towards such big code changes. But we might want to reverse that later on.

**Review by holgerd77 — APPROVED** [review]
Thanks @ryanio, this looks fantastic! 👍  One comment to note, will merge.

**Comment by holgerd77** [comment]
CI run on `master` was [failing](https://github.com/ethereumjs/ethereumjs-client/actions/runs/359796107) after this has been merged, this should be investigated.

**Pull request merged**
